### PR TITLE
Preliminary mixed-size support

### PR DESCRIPTION
**Pull request by mara004**
I've tried to experiment with mixed-size pages. It is a starting point but doesn't work well, because sometimes content is cropped off and I haven't found the cause yet.

### CODE REVIEW

**Comment by mara004** [comment]
An example for which it worked: [out04.pdf](https://github.com/cfcurtis/pdfstitcher/files/6198754/out04.pdf)
An example for which didn't work: [testout.pdf](https://github.com/cfcurtis/pdfstitcher/files/6198759/testout.pdf)

**Comment by cfcurtis** [comment]
I actually had intended to do mixed page support initially but found similar bugginess. I got rid of it a while ago and there are likely other bugs, but [this](https://github.com/cfcurtis/pdfstitcher/blob/2dc419724f0e98ae303eb50c53d49bbbba859182/tile_pages.py) looks like the latest version of the code that attempted to accommodate mixed page sizes if it's of any use to you.

**Comment by mara004** [comment]
Interesting, thanks. I'll compare and try it soon. How did you avoid these crop bugs? Is there something specific one needs to be careful about?

**Comment by cfcurtis** [comment]
I got rid of some crop bugs by setting the TrimBox equal to the MediaBox, but that was for some specific PDFs where they differed (yet displayed correctly in the original PDF, I'm not really sure how that happened).

I took a look at your implementation and I have a couple of comments. First, I would avoid parsing the content stream in this function, as it can be extremely slow with some PDFs, and I also wouldn't count on the first command being the initial matrix. It's unavoidable for layer manipulation, but for page tiling I think you should be able to accomplish everything you want (scaling/shifting/rotating of the page) by reading the various geometry boxes and modifying the transformation matrix when the form xobject is placed.

Secondly, can I ask what your use case is for this? For sewing patterns it's very important that the pages not be scaled, as we want 1" to stay 1" in the final PDF.

**Comment by mara004** [comment]
Okay, I see. It certainly was an exceedingly inelegant implementation with lots of problems, which is because I barely know about PDF internals (I only did arraning/rotating pages so far). Probably I'd better not experiment with this part of the program... I basically just copied the example from [this page](https://pikepdf.readthedocs.io/en/latest/topics/content_streams.html) of the pikepdf documentation.
It's better to continue with your initial implementation and discard this very improper one.

> Secondly, can I ask what your use case is for this? For sewing patterns it's very important that the pages not be scaled, as we want 1" to stay 1" in the final PDF.

The use case is to create overview pages for mixed-size documents.

**Comment by mara004** [comment]
I have now tried out your version and it basically works, but does not really do what I want - for my use case it would be needed that all tilling page boxes are of the same size, that's why I downscaled/upscaled differing pages.

**Comment by mara004** [comment]
I have now tried out your version and it works (performance was indeed significantly better for all documents I tried), but it does not really do what I want - for my use case it would be needed that all tilling page boxes are of the same size, that's why I downscaled/upscaled differing pages...

**Comment by cfcurtis** [comment]
Ah, your target page size and scaling makes more sense to me now. I'm hesitant to merge in changes that add those things as it could result in invalid sewing patterns (my primary use case), but if it's just in this class and not exposed in GUI it'd probably be okay.

The PDF spec is certainly not straightforward. The page tiling function basically packages up each page into a single object (using pikepdf's as_form_xobject) and then places them on a new larger page with the commands:

```
content_txt += f'q {R[0]} {R[1]} {R[2]} {R[3]} {x0} {y0} cm '
content_txt += f'{page_names[i]} Do Q '
```

where the 6 digits before the "cm" operator define the transformation matrix applied to the page xobject. PDFs use 6 digits to define the transformation matrix instead of 9 because the last column is always [0,0,1], so my 6 operands map to:

```
R[0] R[1] 0
R[2] R[3] 0
x0 x1 1
```

To apply scaling you can multiply this transformation matrix with a scaling matrix, e.g.:

```
Sx 0 0
0 Sy 0
0 0 1
```

which is probably easier done with pikepdf's helper matrix class:

`T = pikepdf.PdfMatrix(R + [x0,y0]).scaled(Sx,Sy)
`

which you could then parse back out into the content_txt string. However, since it's matrix multiplication the order of operations matter, and PDFs modify the coordinate system rather than the object, so I might have gotten it backwards.

**Comment by mara004** [comment]
Thanks for explaining, now it begins to dawn on me how that content_txt works - I honestly did not quite understand it before.

**Comment by mara004** [comment]
I've looked into this again this morning and now have created an experimental version that supports mixed pages by either preserving the absolute size and adapting the row height (based on your work) or by target size and scaling to fit. Unless a target size is set, nothing should be scaled. I also added a warning concerning scaling and sewing patterns. It *seems* to work so far for the tests I did. Could you please take a look? Would it theoretically be posisble to merge this if I yet cleaned the code up a bit (remove the debugging prints etc.)?

[tile_pages_new.zip](https://github.com/cfcurtis/pdfstitcher/files/6263647/tile_pages_new.zip)

File: tile_pages.py
Comment (cfcurtis) [inline tile_pages.py:18]: I think the _cpphelpers was added to make it play nicely with pyinstaller, so I think I need to keep it

File: tile_pages.py
Comment (cfcurtis) [inline tile_pages.py:27]: I like this addition, much better for readability

File: tile_pages.py
Comment (cfcurtis) [inline tile_pages.py:321]: f strings don't play nicely with gettext, so this needs to be either the old style .format strings or hardcoded as before.

File: tile_pages.py
Comment (cfcurtis) [inline tile_pages.py:390]: I don't think this will work with rotation - for example if rotation is clockwise then R = [0, -1, 1 0], but with scale_factor = 1 this would write [1, -1, 1, 1] instead. I think you need to do the matrix multiplication between the rotation matrix and scaling matrix.

**Review by cfcurtis — CHANGES_REQUESTED** [review]
Hi, I read through your changes and left a few comments. I haven't tested it yet, but I think you're going to run into issues with page rotation because the matrix multiplication isn't quite right. Otherwise, I think it makes sense to keep working on this to make it general purpose beyond just sewing patterns, and I appreciate your contribution.

File: tile_pages.py
Comment (mara004) [reply tile_pages.py:18]: Ah, I see. I was a bit puzzled why you import _cpphelpers since it isn't used anywhere in this file, but now it's clear it is needed for the installer. I'll revert that.

File: tile_pages.py
Comment (mara004) [reply tile_pages.py:321]: Okay, I think I'll change it to .format()

File: tile_pages.py
Comment (mara004) [reply tile_pages.py:390]: Yes, that's a big mistake. Files indeed look completely wrong when setting rotation. I need to change this.

File: tile_pages.py
Comment (mara004) [inline tile_pages.py:394]: This warning needs to use gettext, I need to wrap it with _()

File: tile_pages.py
Comment (mara004) [reply tile_pages.py:390]: Hmm, I attemted to address the problem, but there's still a shift in y direction that I don't know how to compensate for.

Comment (mara004) [reply tile_pages.py]: It should be fixed now I hope.

**Comment by mara004** [comment]
Not implemented yet:
- center_content without target size
- work with a single target size (width or height)

**Comment by cfcurtis** [comment]
Sorry for the delay, I haven't had much time to test this lately. I'm thinking it would be a good idea to create a new branch and work on these changes before merging them in to main - I've got some translation files coming soon that I want to publish as a new release without changes to the base functionality. I think you need to re-request it though, is that okay? I'll create a new branch called "tile_pages_dev" that you can target for the pull request.

**Comment by mara004** [comment]
No problem, take your time. Yeah, creating a new branch is a good idea. So I will close this pull request and make a new one targetting the tile_pages_dev branch. I only tested the PageTiler class with a few custom scripts, but not the PDFStitcher GUI so far.

**Pull request closed without merging**
